db/dbformat.h

```diff
@@ -160,10 +160,15 @@ class InternalKeyComparator
   }
   virtual ~InternalKeyComparator() {}
 
+  virtual Slice decode_key(const Slice& key) const {
+    return ExtractUserKey(key);
```

Comment:
this doesn't really fit to me. The user key isn't really a decoded version of the internal key. It also makes the `CompareKeySeqDecoded` a bit hard to read because it looks like it reads past the end of `bkey`'s buffer. If we simply return `key` here and fixup `CompareKeySeqDecoded` do you still achieve similar perf improvement?

Reply:
IIRC the avoidance of extra calls to `ExtractUserKey` is a tiny optimization. Fortunately, it's also decoupled from the main change, so it can be separated and sent separately to not block this PR. I'll rework the stuff this evening. :-)